src/tools/rbd_mirror/ImageAction.h

```diff
@@ -0,0 +1,124 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```

Comment:
I almost wonder if ```ImageAction``` should be integrated with / semi-merged into ```Policy``` since right now the in-memory policy map might not actually reflect the reality of where the current state mapping is actually applied. I think that would help eliminate some of the potential race conditions / edge conditions between these two separate image state tracking mechanisms.

Reply:
`ImageAction` is merely just a state machine invoking appropriate context callbacks provided by its consumer. There's no tracking at all in terms of which instance id to is being notified for a given image and the current sub-action being executed. Did you mean parts of `ImageMap` (especially the retry logic in there) to be better integrated with `Policy`?

Reply:
The policy and the state machine should be integrated together so they cannot get out-of-sync. If the policy wants to move an image to a new instance, the policy should be internally tracking the state changes of the image. Right now the in-memory policy can think and image is on instance XYZ while it actually might be still on ABC due to the state machine transition.

Reply:
thanks! I did most of the changes integrating `ImageAction` and `Policy`, but then thought if you meant (by mistake) to move pieces from `ImageMap` to `Policy`. Once we start tracking state changes in `Policy` the retry logic can go away from `ImageMap`.